File: ios/Stopwatch/domain/services/impl/TimerServiceImpl.swift
import Foundation

class TimerServiceImpl<EE>: TimerService where EE: EventEmitter, EE.Event == TimerState {
    private(set) var state: TimerState = TimerState()
    private(set) var events: EE
    
    private var task: Task<(), Never>? = nil
    
    init(_ eventEmitter: EE) {
        self.events = eventEmitter
    }
    
    func start() {
        guard !state.isRunning else { return }
        
        let startMilliseconds = state.milliseconds
        let startEpoch = Date().epochMilliseconds
        
        update(TimerState(isRunning: true, milliseconds: state.milliseconds))
        task = Task {
            while (state.isRunning) {
                do {
                    try await Task.sleep(nanoseconds: 10_000_000)
                    let milliseconds = Date().epochMilliseconds - startEpoch + startMilliseconds
                    update(TimerState(isRunning: state.isRunning, milliseconds: milliseconds))
                } catch {
                    break
                }
            }
        }
    }
    
    func pause() {
        guard state.isRunning else { return }
        
        task?.cancel()
        state = TimerState(isRunning: false, milliseconds: state.milliseconds)
    }
    
    func reset() {
        guard !state.isRunning else { return }
        
        task?.cancel()
        state = TimerState(isRunning: false, milliseconds: 0)
    }
    
    private func update(_ newState: TimerState) {
        state = newState
        events.emit(newState)
    }
}

private extension Date {
    var epochMilliseconds: Int {
        Int(timeIntervalSince1970 * 1_000)
    }
}

File: android/app/src/main/java/com/rafaeltmbr/stopwatch/core/services/impl/TimerServiceImpl.kt
package com.rafaeltmbr.stopwatch.core.services.impl

import com.rafaeltmbr.stopwatch.core.services.TimerService
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class TimerServiceImpl(
    private val coroutineScope: CoroutineScope = CoroutineScope(Dispatchers.Default),
    private val currentMillisecondsCallback: () -> Long = System::currentTimeMillis
) : TimerService {
    private val _state = MutableStateFlow(TimerService.State(isRunning = false, milliseconds = 0L))
    private var timerJob: Job? = null

    override val state: StateFlow<TimerService.State> = _state.asStateFlow()

    override fun set(state: TimerService.State) {
        _state.update { state }

        if (state.isRunning) {
            timerLoop()
        }
    }

    override fun start() {
        if (_state.value.isRunning) return

        _state.update { it.copy(isRunning = true) }
        timerLoop()
    }

    override fun pause() {
        if (!_state.value.isRunning) return

        timerJob?.cancel()
        timerJob = null
        _state.update { it.copy(isRunning = false) }
    }

    override fun reset() {
        if (_state.value.isRunning) return

        _state.update { TimerService.State(isRunning = false, milliseconds = 0L) }
    }

    private fun timerLoop() {
        val startMilliseconds = _state.value.milliseconds
        val startEpoch = currentMillisecondsCallback()

        timerJob?.cancel()
        timerJob = coroutineScope.launch {
            while (_state.value.isRunning) {
                delay(10)
                val milliseconds = currentMillisecondsCallback() - startEpoch + startMilliseconds
                _state.update { it.copy(milliseconds = milliseconds) }
            }
        }
    }
}